lib/helpers.c

```diff
@@ -46,6 +52,28 @@ static void on_connect(struct mosquitto *mosq, void *obj, int rc)
 	mosquitto_subscribe(mosq, NULL, userdata->topic, userdata->qos);
 }
 
+static void on_connect_publish(struct mosquitto *mosq, void *obj, int rc)
+{
+	struct userdata__publish_callback *userdata = obj;
+	struct mosquitto_message *current_msg = NULL;
+	int result;
+	int i;
+
+	for (i=0; i < userdata->msg_count; i++){
+		current_msg = userdata->messages[i];
+
+		result = mosquitto_publish(
+				mosq, NULL, current_msg->topic,
+				current_msg->payloadlen, current_msg->payload,
+				current_msg->qos, current_msg->retain);
+
+		if(result) {
+			break;
+		}
+	}
+
+	mosquitto_disconnect(mosq);
```

Comment:
I'm not sure I like all of this happening in the connect callback, in particular the disconnect call. Could you change it so that the disconnect is called only after the messages are published? Maybe keeping a count of published messages in an on_publish() callback would be the easiest way, then calling disconnect there once the count is reached.